libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp

```diff
@@ -0,0 +1,205 @@
+        // just initialized
+        Debug(1, "LDLidar NO DATA, initializing");
+        set_status(AP_Proximity::Status::NoData);
+        return;
+    }
+
+    // read data
+    read_sensor_data();
+
+    if (AP_HAL::millis() - _last_distance_received_ms < LDROBOT_TIMEOUT_MS) {
+        set_status(AP_Proximity::Status::Good);
+    } else {
+        // long time since we received any valid sensor data
+        // try sending the sensor the "send data" message
+        _initialized = false;
+        set_status(AP_Proximity::Status::NoData);
+        Debug(1, "LDLidar NO DATA");
+    }
+}
+
+
+void AP_Proximity_LDLidar_STL27L::read_sensor_data()
+{
+    uint32_t nbytes = _uart->available();
+    if(nbytes > 0)
+    {
+        Debug(1, "LDLidar got %ld bytes data", nbytes);
+    }
+    while (nbytes-- > 0) {
+        int16_t byte = _uart->read();
+        if (!parse_byte(byte)) {
+            // reset
+            reset();
+        }
+    }
+}
+
+// parse one byte from the sensor. Return false on error.
+bool AP_Proximity_LDLidar_STL27L::parse_byte(uint8_t data)
+{
+    switch (_parse_state) {
+    case Header:
+        if (data == LDROBOT_PACKET_HEADER) {
+            _packet._frame.header = data;
+            _parse_state = VerLen;
+            _bytes_count++;
+            Debug(1, "LDLidar Found Header");
+            return true;
+        }
+
+        return false;
+
+    case VerLen:
+        if (data == LDROBOT_PACKET_VER_LEN) {
+            _packet._frame.ver_len = data;
+            _parse_state = Data;
+            _bytes_count++;
+            Debug(1, "LDLidar Found VerLen");
+            return true;
+        }
+        //ver_len error
+        return false;
+
+    case Data:
+        //1. copy data
+        //2. count++
+        _packet.data[_bytes_count++] = data;
+        //3. if count >= size, a) verify checksum, b)parse packet and update distance, c)reset()
+        if (_bytes_count == LDROBOT_PACKET_TOTAL_LEN) {
+            Debug(1, "LDLidar Found Packet");
+            const uint8_t check_sum = calc_checksum(_packet.data, LDROBOT_PACKET_TOTAL_LEN - 1);
+            if(check_sum != data)
+            {
+                Debug(1, "LDLidar drop Packet, checksum error");
+                return false;
+            }
+            //parse packet
+            _last_distance_received_ms = AP_HAL::millis();
+            parse_payload();
+            //update distance
+            _temp_boundary.update_3D_boundary(state.instance, frontend.boundary);
+            reset();
+        }
+        return true;
+
+    default:
+        return false;
+    }
+
+    return false;
+}
+
+// parse payload, to pick out distances, and feed them to the correct faces
+void AP_Proximity_LDLidar_STL27L::parse_payload()
+{
+    uint32_t diff = ((uint32_t)_packet._frame.end_angle + 36000 - (uint32_t)_packet._frame.start_angle) % 36000;
+    float step = diff / (LDROBOT_POINT_PER_PACK - 1) / 100.0;
```

Comment:
```suggestion
    const float step = diff / (LDROBOT_POINT_PER_PACK - 1) * 0.01;
```Similarly for the lines below